File: ios/ClaudeScope/Sources/Models/StoredCredentials.swift
import Foundation

struct StoredCredentials: Codable, Equatable {
    let accessToken: String
    let refreshToken: String?
    let expiresAt: Date?
    let scopes: [String]

    var hasRefreshToken: Bool {
        guard let refreshToken else { return false }
        return !refreshToken.isEmpty
    }

    func needsRefresh(at now: Date = Date(), leeway: TimeInterval = 300) -> Bool {
        guard hasRefreshToken, let expiresAt else { return false }
        return expiresAt <= now.addingTimeInterval(leeway)
    }

    func isExpired(at now: Date = Date()) -> Bool {
        guard let expiresAt else { return false }
        return expiresAt <= now
    }
}

struct StoredCredentialsStore {
    private let fileManager: FileManager
    let directoryURL: URL
    let credentialsFileURL: URL

    init(fileManager: FileManager = .default) {
        self.fileManager = fileManager
        let appSupport = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        self.directoryURL = appSupport.appendingPathComponent("ClaudeScope", isDirectory: true)
        self.credentialsFileURL = directoryURL.appendingPathComponent("credentials.json")
    }

    func save(_ credentials: StoredCredentials) throws {
        try ensureDirectoryExists()
        let data = try Self.encoder.encode(credentials)
        try data.write(to: credentialsFileURL, options: .atomic)
    }

    func load(defaultScopes: [String]) -> StoredCredentials? {
        guard let data = try? Data(contentsOf: credentialsFileURL),
              let credentials = try? Self.decoder.decode(StoredCredentials.self, from: data) else {
            return nil
        }
        return credentials
    }

    func delete() {
        try? fileManager.removeItem(at: credentialsFileURL)
    }

    private func ensureDirectoryExists() throws {
        try fileManager.createDirectory(at: directoryURL, withIntermediateDirectories: true)
    }

    private static let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        return encoder
    }()

    private static let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }()
}

File: android/app/src/main/kotlin/io/sandwichlab/claudescope/data/model/StoredCredentials.kt
package io.sandwichlab.claudescope.data.model

data class StoredCredentials(
    val accessToken: String,
    val refreshToken: String?,
    val expiresAtEpochMs: Long?,
    val scopes: List<String>,
) {
    val hasRefreshToken: Boolean
        get() = !refreshToken.isNullOrEmpty()

    fun isExpired(now: Long = System.currentTimeMillis()): Boolean =
        expiresAtEpochMs != null && expiresAtEpochMs <= now

    fun needsRefresh(
        now: Long = System.currentTimeMillis(),
        leewayMs: Long = REFRESH_LEEWAY_MS,
    ): Boolean {
        if (!hasRefreshToken) return false
        val expires = expiresAtEpochMs ?: return false
        return expires <= now + leewayMs
    }

    companion object {
        const val REFRESH_LEEWAY_MS: Long = 5 * 60 * 1000L
    }
}
